TestMissedRevenueHeatMapGenerator.py

```diff
@@ -0,0 +1,187 @@
+        
+        self.assertEqual("calculate_stock_level() returned incorrect value", calculated_stock_level, expected_stock_level)
+
+    def test_calculate_low_stock_level_periods(self):
+        '''Calculate the periods in time for which stock level was below the danger level.
+           The danger level is defined as the stock level at which customers don’t buy.
+           The danger period is defined as a period of time where stock level is equal or less than danger level.'''
+        #define a journal to simulate low stock level periods
+        journal = [{'Branch':self.branch, 'Date':self.time_branch_opens, 'Increment':10, 'Sku':self.sku, 'Tag':'purchase'},
+                   {'Branch':self.branch, 'Date':self.sale_date_1, 'Increment':-10, 'Sku':self.sku, 'Tag':'sale'},
+                   {'Branch':self.branch, 'Date':self.purchase_date_1, 'Increment':1, 'Sku':self.sku, 'Tag':'purchase'},
+                   {'Branch':self.branch, 'Date':self.sale_date_2, 'Increment':-1, 'Sku':self.sku, 'Tag':'sale'},
+                   {'Branch':self.branch, 'Date':self.purchase_date_2, 'Increment':20, 'Sku':self.sku, 'Tag':'purchase'}]
+        #there are two periods during which stock is 0
+        danger_period_1 = (self.sale_date_1, self.purchase_date_1)
+        danger_period_2 = (self.sale_date_2, self.purchase_date_2)
+        
+        #return a tuple with time periods when stock was zero
+        calculated_danger_period = MR.calculate_low_stock_level_periods(self.sku, self.branch, journal)
+
+        self.assertEqual("calculated_danger_period() returned incorrect value", calculated_danger_period, (danger_period_1, danger_period_2))
+
+        #test with danger_level = 1
+        #TODO: Create seperate tests for varying danger levels
+        danger_period_3 = (sale_date_1, purchase_date_2)
+        calculated_danger_period = MR.calculate_low_stock_level_periods(self.sku, self.branch, journal, 1)
+
+        self.assertEqual("calculate_low_stock_level_periods() returned incorrect value", calculated_danger_period, (danger_period_3,))
+        
+    #TODO: Add test to ensure that low_stock_level_periods is always within business hours
+        
+    def test_calculate_average_rate_of_sale_per_hour(self):
+        '''Test the calculated average hourly rate of sale for one day'''
+        #TODO: Calculate average rate for larger periods
+        #reuse calculate_hours_a_branch_is_open function - calculate hours open during a specific day
+        #3 Jan 2017 was a Tuesday
+        hours_open = 8
+        journal = [{'Branch':self.branch, 'Date':self.sale_date_1, 'Increment':-6, 'Sku':self.sku, 'Tag':'sale'},
+                   {'Branch':self.branch, 'Date':self.purchase_date_1, 'Increment':2, 'Sku':self.sku, 'Tag':'purchase'},
+                   {'Branch':self.branch, 'Date':self.sale_date_2, 'Increment':-7, 'Sku':self.sku, 'Tag':'sale'}]
+        
+        #increment of sales is negative
+        #sales = 6 + 7
+        sales = 13
```

Comment:
it is used in the calculate_average_rate_of_sale_per_hour method.
calculate_average_rate_of_sale_per_hour(self, sku, branch, journal, hours_open)